### ISSUE TITLE
The Conditional Random Field is giving me a negative loss

### ISSUE THREAD

**Issue by Sleemanmunk** [allenai/allennlp#2658]
**System (please complete the following information):**
 - OS: Linux
 - Python version: 3.6.8
 - AllenNLP version: v0.8.2
 - PyTorch version: 1.0.0

**Question**
I am building my own NER tool using your conditional random field. In general, it all works splendidly, but occasionally, I find that my cost goes negative. I did not have this problem just doing softmax-based sequence labeling. Since the cost is calculated as -1 times the log joint likelihood minus the log input likelihood, negative cost suggests the input is less likely than the input _and_ the underlying distribution, which is impossible. I am aware that the CRF itself does not apply the (*-1) and have applied it myself.

I know that doesn't mean that some small approximation error couldn't cause this behavior, but needless to say, it breaks my model. Is this expected behavior in edge cases where the model has drastically overfit? Should I just apply more regularization?

I've made sure that I'm using the mask properly (ones for the length of each sentence, followed by zeros for padding).

**Comment by vidurj** [comment]
@Sleemanmunk could you provide us with the `inputs`, `tags` and `mask` arguments to `ConditionalRandomField.forward` that result in a positive output?

**Comment by vidurj** [comment]
Thank you for those! PR #2680 should fix this. Your `tags` and `mask` tensor are longer than `inputs` in the time dimension (42 versus 113). Is this intended? If you make them the same length, it will work with the module you are currently using.

**Comment by Sleemanmunk** [comment]
I found and fixed the problem where they were different lengths, but that didn't seem to solve the issue.
I am still getting a positive output. Here are my new files.
[sequences_same_length.zip](https://github.com/allenai/allennlp/files/3035159/sequences_same_length.zip)

**Comment by vidurj** [comment]
This is a numerical precision issue. You could get around it by making the CRF model and the input `logits` double precision, or ignoring the issue.

**Comment by vidurj** [comment]
To summarize, 
1) AllenNLP assumes that the mask has the right shape for its inputs.
2) The CRF model can output slightly positive values because of precision issues.
